### PR TITLE
Support the `http.request.method_original` attribute

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Implements https://github.com/open-telemetry/semantic-conventions/pull/17

### CODE REVIEW

File: instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpServerAttributesExtractor.java [hunk 1]
```diff
@@ -78,12 +79,14 @@ public static <REQUEST, RESPONSE> HttpServerAttributesExtractorBuilder<REQUEST,
...
-      List<String> capturedResponseHeaders) {
+      List<String> capturedResponseHeaders,
+      Set<String> knownMethods) {
     this(
         httpAttributesGetter,
         netAttributesGetter,
         capturedRequestHeaders,
         capturedResponseHeaders,
+        knownMethods,
         HttpRouteHolder::getRoute);
```
Comment (mateuszrzeszutek) [inline instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpServerAttributesExtractor.java:90]: The complexity of these constructors is kinda starting to get out of hand; I'm planning to implement Jason's dependency injection comment once I merge all the in-flight PRs.

File: instrumentation/okhttp/okhttp-3.0/library/src/main/java/io/opentelemetry/instrumentation/okhttp/v3_0/OkHttpTelemetryBuilder.java [hunk 2]
```diff
@@ -24,6 +26,7 @@ public final class OkHttpTelemetryBuilder {
       new ArrayList<>();
   private List<String> capturedRequestHeaders = emptyList();
   private List<String> capturedResponseHeaders = emptyList();
+  @Nullable private Set<String> knownMethods = null;
```
Comment (mateuszrzeszutek) [inline instrumentation/okhttp/okhttp-3.0/library/src/main/java/io/opentelemetry/instrumentation/okhttp/v3_0/OkHttpTelemetryBuilder.java:29]: I really dislike these `null`s here; but this was the least invasive way of implementing an optionally unset set of methods. I've an idea on how to refactor these, but I'll tackle that in a separate PR.

File: instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpCommonAttributesExtractor.java [hunk 3]
```diff
@@ -35,19 +38,29 @@ abstract class HttpCommonAttributesExtractor<
...
 
   @Override
   public void onStart(AttributesBuilder attributes, Context parentContext, REQUEST request) {
     String method = getter.getHttpRequestMethod(request);
     if (SemconvStability.emitStableHttpSemconv()) {
-      internalSet(attributes, HttpAttributes.HTTP_REQUEST_METHOD, method);
+      if (knownMethods.contains(method)) {
+        internalSet(attributes, HttpAttributes.HTTP_REQUEST_METHOD, method);
+      } else {
+        internalSet(attributes, HttpAttributes.HTTP_REQUEST_METHOD, _OTHER);
```
Comment (breedx-splk) [inline instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpCommonAttributesExtractor.java:61]: static import the names to improve readability?

File: instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpServerAttributesExtractor.java (hunk 1 above)
Comment (breedx-splk) [reply instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpServerAttributesExtractor.java:90]: Because this constructor is called all over the place with the default set, you might (in the short term) also consider doing a constructor overload that defaults the set. I think that applies to the client attr extractor as well.

File: instrumentation-api-semconv/src/testStableHttpSemconv/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpClientAttributesExtractorStableSemconvTest.java [hunk 4]
```diff
@@ -219,4 +222,82 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
...
+
+    AttributesBuilder attributes = Attributes.builder();
+    extractor.onStart(attributes, Context.root(), request);
+    extractor.onEnd(attributes, Context.root(), request, emptyMap(), null);
+
+    assertThat(attributes.build()).containsEntry(HttpAttributes.HTTP_REQUEST_METHOD, requestMethod);
+  }
+
+  @ParameterizedTest
+  @ValueSource(strings = {"get", "PURGE", "Get", "not a method really"})
```
Comment (breedx-splk) [inline instrumentation-api-semconv/src/testStableHttpSemconv/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpClientAttributesExtractorStableSemconvTest.java:245]: Nit: This is kinda mixing the "not-a-method" testing with "methods are case sensitive". I'd split the case sensitive stuff into its own test.

File: instrumentation-api-semconv/src/testStableHttpSemconv/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpClientAttributesExtractorStableSemconvTest.java [hunk 5]
```diff
@@ -219,4 +222,82 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
...
+    AttributesBuilder attributes = Attributes.builder();
+    extractor.onStart(attributes, Context.root(), request);
+    extractor.onEnd(attributes, Context.root(), request, emptyMap(), null);
+
+    assertThat(attributes.build()).containsEntry(HttpAttributes.HTTP_REQUEST_METHOD, requestMethod);
+  }
+
+  @ParameterizedTest
+  @ValueSource(
+      strings = {"CONNECT", "DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT", "TRACE"})
```
Comment (breedx-splk) [inline instrumentation-api-semconv/src/testStableHttpSemconv/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpClientAttributesExtractorStableSemconvTest.java:284]: Nit: Might be nice if these values came from the constant set instead of being duplicated. Maybe a `@MethodSource` that returns those could help.

File: instrumentation-api-semconv/src/testStableHttpSemconv/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpServerAttributesExtractorStableSemconvTest.java [hunk 6]
```diff
@@ -243,4 +246,82 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
...
+
+    AttributesExtractor<Map<String, String>, Map<String, String>> extractor =
+        HttpServerAttributesExtractor.create(
+            new TestHttpServerAttributesGetter(), new TestNetServerAttributesGetter());
+
+    AttributesBuilder attributes = Attributes.builder();
+    extractor.onStart(attributes, Context.root(), request);
+    extractor.onEnd(attributes, Context.root(), request, emptyMap(), null);
+
+    assertThat(attributes.build()).containsEntry(HttpAttributes.HTTP_REQUEST_METHOD, requestMethod);
```
Comment (breedx-splk) [inline instrumentation-api-semconv/src/testStableHttpSemconv/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpServerAttributesExtractorStableSemconvTest.java:265]: Could also verify that the original method isn't set. 🤷🏻

File: instrumentation-api-semconv/src/testStableHttpSemconv/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpServerAttributesExtractorStableSemconvTest.java [hunk 7]
```diff
@@ -243,4 +246,82 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
...
+  }
+
+  @ParameterizedTest
+  @ValueSource(
+      strings = {"CONNECT", "DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT", "TRACE"})
```
Comment (breedx-splk) [inline instrumentation-api-semconv/src/testStableHttpSemconv/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpServerAttributesExtractorStableSemconvTest.java:308]: Same comment as above about using the original/actual constant set.

File: instrumentation/java-http-client/library/src/main/java/io/opentelemetry/instrumentation/httpclient/JavaHttpClientTelemetryBuilder.java [hunk 8]
```diff
@@ -65,10 +68,32 @@ public JavaHttpClientTelemetryBuilder setCapturedResponseHeaders(List<String> re
...
+   * method defined in <a href="https://www.rfc-editor.org/rfc/rfc5789.html">RFC5789</a>.
+   *
+   * <p>Note: calling this method <b>overrides</b> the default known method sets completely; it does
+   * not supplement it.
+   *
+   * @param knownMethods A set of recognized HTTP request methods.
+   */
+  @CanIgnoreReturnValue
+  public JavaHttpClientTelemetryBuilder setKnownMethods(Set<String> knownMethods) {
+    this.knownMethods = knownMethods;
```
Comment (breedx-splk) [inline instrumentation/java-http-client/library/src/main/java/io/opentelemetry/instrumentation/httpclient/JavaHttpClientTelemetryBuilder.java:85]: Nit: Wrap with a new set so the caller can't sneaky mutate later. Arguably overkill for a builder.

File: instrumentation/ktor/ktor-1.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v1_0/KtorServerTracing.kt [hunk 9]
```diff
@@ -75,6 +75,10 @@ class KtorServerTracing private constructor(
       httpAttributesExtractorBuilder.setCapturedResponseHeaders(responseHeaders)
     }
 
+    fun setKnownMethods(knownMethods: Set<String>) {
+      httpAttributesExtractorBuilder.setKnownMethods(knownMethods)
```
Comment (breedx-splk) [inline instrumentation/ktor/ktor-1.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v1_0/KtorServerTracing.kt:79]: I think it might be more idiomatic in kotlin to do this (assuming it compiles, lol):
```suggestion
      httpAttributesExtractorBuilder.knownMethods = knownMethods
```

File: instrumentation/ktor/ktor-2.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/client/KtorClientTracingBuilder.kt [hunk 10]
```diff
@@ -44,6 +44,10 @@ class KtorClientTracingBuilder {
     httpAttributesExtractorBuilder.setCapturedResponseHeaders(headers)
   }
 
+  fun setKnownMethods(knownMethods: Set<String>) {
+    httpAttributesExtractorBuilder.setKnownMethods(knownMethods)
```
Comment (breedx-splk) [inline instrumentation/ktor/ktor-2.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/client/KtorClientTracingBuilder.kt:48]:
```suggestion
    httpAttributesExtractorBuilder.knownMethods = knownMethods
```

File: instrumentation/ktor/ktor-2.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/server/KtorServerTracing.kt [hunk 11]
```diff
@@ -76,6 +76,10 @@ class KtorServerTracing private constructor(
       httpAttributesExtractorBuilder.setCapturedResponseHeaders(responseHeaders)
     }
 
+    fun setKnownMethods(knownMethods: Set<String>) {
+      httpAttributesExtractorBuilder.setKnownMethods(knownMethods)
```
Comment (breedx-splk) [inline instrumentation/ktor/ktor-2.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/server/KtorServerTracing.kt:80]:
```suggestion
      httpAttributesExtractorBuilder.knownMethods = knownMethods
```

File: instrumentation/netty/netty-4-common/library/src/main/java/io/opentelemetry/instrumentation/netty/v4/common/internal/server/NettyServerInstrumenterFactory.java [hunk 12]
```diff
@@ -27,19 +30,23 @@ public static Instrumenter<HttpRequestAndChannel, HttpResponse> create(
...
     NettyHttpServerAttributesGetter httpAttributesGetter = new NettyHttpServerAttributesGetter();
 
+    HttpServerAttributesExtractorBuilder<HttpRequestAndChannel, HttpResponse> extractorBuilder =
+        HttpServerAttributesExtractor.builder(
+                httpAttributesGetter, new NettyNetServerAttributesGetter())
+            .setCapturedRequestHeaders(capturedRequestHeaders)
+            .setCapturedResponseHeaders(capturedResponseHeaders);
+    if (knownMethods != null) {
+      extractorBuilder.setKnownMethods(knownMethods);
+    }
```
Comment (breedx-splk) [inline instrumentation/netty/netty-4-common/library/src/main/java/io/opentelemetry/instrumentation/netty/v4/common/internal/server/NettyServerInstrumenterFactory.java:45]: If the builder internally guarded against `null` and just did a noop, these multiple`if` checks sprinkled about could go away. Basically encapsulate this null check.

File: instrumentation/okhttp/okhttp-3.0/library/src/main/java/io/opentelemetry/instrumentation/okhttp/v3_0/OkHttpTelemetryBuilder.java [hunk 13]
```diff
@@ -62,13 +65,35 @@ public OkHttpTelemetryBuilder setCapturedResponseHeaders(List<String> responseHe
...
+   * href="https://www.rfc-editor.org/rfc/rfc9110.html#name-methods">RFC9110</a> and the PATCH
+   * method defined in <a href="https://www.rfc-editor.org/rfc/rfc5789.html">RFC5789</a>.
+   *
+   * <p>Note: calling this method <b>overrides</b> the default known method sets completely; it does
+   * not supplement it.
+   *
+   * @param knownMethods A set of recognized HTTP request methods.
+   */
+  @CanIgnoreReturnValue
+  public OkHttpTelemetryBuilder setKnownMethods(Set<String> knownMethods) {
```
Comment (breedx-splk) [inline instrumentation/okhttp/okhttp-3.0/library/src/main/java/io/opentelemetry/instrumentation/okhttp/v3_0/OkHttpTelemetryBuilder.java:81]: `@Nullable` on the param?

File: instrumentation/spring/spring-webflux/spring-webflux-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webflux/v5_3/SpringWebfluxTelemetryBuilder.java [hunk 14]
```diff
@@ -127,6 +128,25 @@ public SpringWebfluxTelemetryBuilder setCapturedServerResponseHeaders(
...
+   *
+   * <p>Note: calling this method <b>overrides</b> the default known method sets completely; it does
+   * not supplement it.
+   *
+   * @param knownMethods A set of recognized HTTP request methods.
+   */
+  @CanIgnoreReturnValue
+  public SpringWebfluxTelemetryBuilder setKnownMethods(Set<String> knownMethods) {
+    clientBuilder.setKnownMethods(knownMethods);
+    httpServerAttributesExtractorBuilder.setKnownMethods(knownMethods);
```
Comment (breedx-splk) [inline instrumentation/spring/spring-webflux/spring-webflux-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webflux/v5_3/SpringWebfluxTelemetryBuilder.java:146]: Not your issue, but I am noticing the inconsistent naming between client/server fields here.

File: javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/bootstrap/internal/CommonConfig.java [hunk 15]
```diff
@@ -5,18 +5,26 @@
...
 
 /**
  * This class is internal and is hence not for public use. Its APIs are unstable and can change at
  * any time.
  */
 public final class CommonConfig {
 
+  private static final List<String> DEFAULT_KNOWN_METHODS =
+      unmodifiableList(
+          asList("CONNECT", "DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT", "TRACE"));
```
Comment (breedx-splk) [inline javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/bootstrap/internal/CommonConfig.java:26]: <grump> I don't like that this set is duplicated here and in `HttpRequestMethodUtil`. It makes it too easy down the road to change one and not the other. Is there common place that these can be located?

**Review by breedx-splk — COMMENTED** [review]
Looking good to me, just had a few small ideas. Thanks!

File: instrumentation/ktor/ktor-1.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v1_0/KtorServerTracing.kt (hunk 9 above)
Comment (mateuszrzeszutek) [reply instrumentation/ktor/ktor-1.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v1_0/KtorServerTracing.kt:79]: Unfortunately it doesn't work 😢 
```
Cannot access 'knownMethods': it is package-private in 'HttpServerAttributesExtractorBuilder'
```

File: instrumentation/netty/netty-4-common/library/src/main/java/io/opentelemetry/instrumentation/netty/v4/common/internal/server/NettyServerInstrumenterFactory.java (hunk 12 above)
Comment (mateuszrzeszutek) [reply instrumentation/netty/netty-4-common/library/src/main/java/io/opentelemetry/instrumentation/netty/v4/common/internal/server/NettyServerInstrumenterFactory.java:45]: I don't really want to do that -- passing `null`s or using `@Nullable`s on collections (or almost any type that has a "natural" null object value) is a code smell to me. I'm planning to get rid of these horrible nullchecks when I implement https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/8779#discussion_r1237096199

File: instrumentation/okhttp/okhttp-3.0/library/src/main/java/io/opentelemetry/instrumentation/okhttp/v3_0/OkHttpTelemetryBuilder.java (hunk 13 above)
Comment (mateuszrzeszutek) [reply instrumentation/okhttp/okhttp-3.0/library/src/main/java/io/opentelemetry/instrumentation/okhttp/v3_0/OkHttpTelemetryBuilder.java:81]: Collections should never be nullable in the API (see  [the comment above](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/8779#discussion_r1241950700))

File: instrumentation/spring/spring-webflux/spring-webflux-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webflux/v5_3/SpringWebfluxTelemetryBuilder.java (hunk 14 above)
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-webflux/spring-webflux-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webflux/v5_3/SpringWebfluxTelemetryBuilder.java:146]: Yeah, that's true -- but it's also because the `clientBuilder` has a completely different type; it's an internal instrumenter builder wrapper.

File: instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpCommonAttributesExtractor.java (hunk 3 above)
Comment (mateuszrzeszutek) [reply instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpCommonAttributesExtractor.java:61]: The `HttpAttributes` values? I'm in favor of that actually -- but if we're doing that, we should implement it for all attribute keys across the whole package. I think I'd prefer to make it a separate PR, if you don't mind

File: instrumentation/java-http-client/library/src/main/java/io/opentelemetry/instrumentation/httpclient/JavaHttpClientTelemetryBuilder.java (hunk 8 above)
Comment (mateuszrzeszutek) [reply instrumentation/java-http-client/library/src/main/java/io/opentelemetry/instrumentation/httpclient/JavaHttpClientTelemetryBuilder.java:85]: This is gonna go away when I implement https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/8779#discussion_r1237096199

File: instrumentation/netty/netty-4-common/library/src/main/java/io/opentelemetry/instrumentation/netty/v4/common/internal/server/NettyServerInstrumenterFactory.java (hunk 12 above)
Comment (breedx-splk) [reply instrumentation/netty/netty-4-common/library/src/main/java/io/opentelemetry/instrumentation/netty/v4/common/internal/server/NettyServerInstrumenterFactory.java:45]: Legit. 👍🏻

File: instrumentation/spring/spring-webflux/spring-webflux-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webflux/v5_3/SpringWebfluxTelemetryBuilder.java (hunk 14 above)
Comment (breedx-splk) [reply instrumentation/spring/spring-webflux/spring-webflux-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webflux/v5_3/SpringWebfluxTelemetryBuilder.java:146]: Yeah ok. 🤷🏻

**Comment by mateuszrzeszutek** [comment]
@open-telemetry/java-instrumentation-approvers please review this PR 🙏

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/HttpConstants.java [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+ */
+public final class HttpConstants {
+
+  public static final Set<String> KNOWN_METHODS =
+      unmodifiableSet(
+          new HashSet<>(
+              asList(
+                  "CONNECT", "DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT", "TRACE")));
+
+  public static final String _OTHER = "_OTHER";
```
Comment (laurit) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/HttpConstants.java:26]: why the `_`?

**Review by laurit — APPROVED** [review]

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/HttpConstants.java (hunk 16 above)
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/HttpConstants.java:26]: It was chosen because it makes the "other" default value look like it's a kinda special thing (which it is)

> If the HTTP request method is not known to instrumentation, it MUST set the `http.request.method` attribute to `_OTHER` and, except if reporting a metric, MUST set the exact method received in the request line as value of the `http.request.method_original` attribute.

Comment (laurit) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/HttpConstants.java:26]: If it comes from the spec then so it must be. In my opinion using `_OTHER` is a mistake, there is no way that this is not going to confuse users. Everybody who sees it will assume that this is bug/typo.

Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/HttpConstants.java:26]: 🤷 
I don't really have a strong opinion on the value itself; I do agree that you need to have some sort of indication that it's some other unrecognized value though.

Comment (trask) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/HttpConstants.java:26]: this is probably the most relevant comment why `_OTHER` was chosen: https://github.com/open-telemetry/semantic-conventions/pull/17#discussion_r1230390256

I could see backends potentially displaying this metric value with a special tooltip to help ease user confusion

File: instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/NettyServerTelemetryBuilder.java [hunk 17]
```diff
@@ -46,13 +49,32 @@ public NettyServerTelemetryBuilder setCapturedResponseHeaders(
...
+   * <p>Note: calling this method <b>overrides</b> the default known method sets completely; it does
+   * not supplement it.
+   *
+   * @param knownMethods A set of recognized HTTP request methods.
+   */
+  @CanIgnoreReturnValue
+  public NettyServerTelemetryBuilder setKnownMethods(Set<String> knownMethods) {
+    this.knownMethods = knownMethods;
+    return this;
+  }
```
Comment (trask) [inline instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/NettyServerTelemetryBuilder.java:68]: does it make any sense to have `HttpClientTelemetryBuilder` interface to help enforce that these implementations are consistent (and maybe we can inherit javadoc)?

**Review by trask — APPROVED** [review]
I'll open an issue to add common http client/server tests which exercise the non-standard methods (I don't think it's a high priority since it's an edge case anyways)

File: instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/NettyServerTelemetryBuilder.java (hunk 17 above)
Comment (mateuszrzeszutek) [reply instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/NettyServerTelemetryBuilder.java:68]: > (and maybe we can inherit javadoc)?

Hmm, we can put a `@see` tag I think. I don't think we can actually enforce that though.
